framework/src/base/MeshGeneratorSystem.C

```diff
@@ -0,0 +1,496 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MeshGeneratorSystem.h"
+
+#include "MooseApp.h"
+#include "MeshGenerator.h"
+#include "DependencyResolver.h"
+
+MeshGeneratorSystem::MeshGeneratorSystem(MooseApp & app)
+  : PerfGraphInterface(app.perfGraph(), "MeshGeneratorSystem"),
+    ParallelObject(app),
+    _app(app),
+    _popped_final_mesh_generator(false)
+{
+}
+
+void
+MeshGeneratorSystem::addMeshGenerator(const std::string & type,
+                                      const std::string & name,
+                                      const InputParameters & params)
+{
+  mooseAssert(!_mesh_generator_params.count(name), "Already exists");
+  _mesh_generator_params.emplace(
+      std::piecewise_construct, std::forward_as_tuple(name), std::forward_as_tuple(type, params));
+
+  // This should be a sub mesh generator. We can assume this because if we're in the middle of
+  // constructing mesh generators (not "adding" them, where we simply store their parameters),
+  // that we're constructing a sub generator also in the correct order
+  if (_app.constructingMeshGenerators())
+    createMeshGenerator(name);
+}
+
+const MeshGenerator &
+MeshGeneratorSystem::appendMeshGenerator(const std::string & type,
+                                         const std::string & name,
+                                         InputParameters params)
+{
+  if (!appendingMeshGenerators())
+    mooseError("Can only call appendMeshGenerator() during the append_mesh_generator task");
+
+  // Make sure this mesh generator has one and _only_ one input, as "input"
+  const auto param_name_mg_name_pairs = getMeshGeneratorParamDependencies(params, true);
+  if (param_name_mg_name_pairs.size() != 1 || param_name_mg_name_pairs[0].first != "input")
+    mooseError("While adding ",
+               type,
+               " '",
+               name,
+               "' via appendMeshGenerator():\nCan only append a mesh generator that takes a "
+               "single input mesh generator via the parameter named 'input'");
+
+  // If no final generator is set, we need to make sure that we have one; we will hit
+  // this the first time we add an appended MeshGenerator and only need to do it once.
+  // We'll then generate the final ordering within the execution phase. We'll also
+  // clear the ordering because it isn't particularly valid anymore once we add this one.
```

Comment:
```suggestion
  // clear the ordering because it isn't particularly valid anymore once we add the final one.
```

Reply:
```diff
diff --git a/framework/src/base/MeshGeneratorSystem.C b/framework/src/base/MeshGeneratorSystem.C
index 5b97fe376a..f4c2df5b93 100644
--- a/framework/src/base/MeshGeneratorSystem.C
+++ b/framework/src/base/MeshGeneratorSystem.C
@@ -58,7 +58,9 @@ MeshGeneratorSystem::appendMeshGenerator(const std::string & type,
   // If no final generator is set, we need to make sure that we have one; we will hit
   // this the first time we add an appended MeshGenerator and only need to do it once.
   // We'll then generate the final ordering within the execution phase. We'll also
-  // clear the ordering because it isn't particularly valid anymore once we add this one.
+  // clear the ordering because it could be invalid if we append any more generators,
+  // and we'll be re-ordering within executeMeshgenerators() anyway (where we don't
+  // keep track of any state for the sake of simpler logic)
   if (_final_generator_name.empty())
   {
     if (_mesh_generators.empty())
````

Reply:
this works